pkg/proxy/backend/backend_conn_mgr.go

```diff
@@ -798,3 +835,101 @@ func (mgr *BackendConnManager) ConnInfo() []zap.Field {
+		return false
+	}
+
+	mgr.signalReceived <- signalTypePause
+	return true
+}
+
+// tryPause read sessionStates from backend and store the states.
+func (mgr *BackendConnManager) tryPause(ctx context.Context) {
+	backendIO := mgr.backendIO.Load()
+	if backendIO == nil {
+		err := errors.New("backend is nil")
+		mgr.logger.Error("failed to get backend when pause", zap.Error(err))
+		return
+	}
+	addr := mgr.ServerAddr()
+	sessionStates, sessionToken, err := mgr.querySessionStates(backendIO)
+	if err != nil {
+		mgr.logger.Error("query session failed when pause", zap.Error(err))
+		mgr.notifyPauseResult(addr, err)
+		return
+	}
+	mgr.sessionState.Store(&SessionState{
+		sessionStates: sessionStates,
+		sessionToken:  sessionToken,
+	})
+	mgr.backendIO.Store(nil)
+	mgr.lastPauseTime = monotime.Now()
+	mgr.notifyPauseResult(addr, nil)
+	if ignoredErr := backendIO.Close(); ignoredErr != nil && !pnet.IsDisconnectError(ignoredErr) {
+		mgr.logger.Error("close paused backend connection failed", zap.Error(ignoredErr))
+	}
+}
+
+// notify pause result to eventReceiver.
+func (mgr *BackendConnManager) notifyPauseResult(addr string, pauseErr error) {
+	eventReceiver := mgr.getEventReceiver()
+	if eventReceiver == nil {
+		return
+	}
+	if pauseErr != nil {
+		err := eventReceiver.OnPauseFail(addr, mgr)
+		mgr.logger.Warn("pause connection failed", zap.String("addr", addr),
+			zap.NamedError("pause_err", pauseErr), zap.NamedError("notify_err", err))
+	} else {
+		err := eventReceiver.OnPauseSucceed(addr, mgr)
+		mgr.logger.Debug("pause connection succeeds", zap.String("from", addr),
+			zap.NamedError("notify_err", err))
+	}
+}
+
+// resume will reconnect to a new backend when the connection is paused.
+// Then restore the saved session to the backend.
+func (mgr *BackendConnManager) resume(ctx context.Context) error {
+	token := new(SessionToken)
+	state := mgr.sessionState.Load()
+	if state == nil {
+		err := errors.New("session state is nil")
+		return err
+	}
+	err := json.Unmarshal([]byte(state.sessionToken), token)
+	if err != nil {
+		return err
+	}
+
+	clientResp := pnet.HandshakeResp{User: token.Username}
+	mgr.logger.Info("start to resume session")
+	newBackendIO, err := mgr.getBackendIO(ctx, mgr, &clientResp)
+	if err != nil {
+		return err
+	}
+	mgr.logger.Debug("start to handshakeSecondTime")
+	if err = mgr.authenticator.handshakeSecondTime(mgr.logger, mgr.clientIO, newBackendIO, mgr.backendTLS, state.sessionToken); err == nil {
+		err = mgr.initSessionStates(newBackendIO, state.sessionStates)
+	} else {
+		mgr.handshakeHandler.OnHandshake(mgr, newBackendIO.RemoteAddr().String(), err, Error2Source(err))
+	}
+	mgr.logger.Debug("finish to handshakeSecondTime")
+	if err != nil {
+		if ignoredErr := newBackendIO.Close(); ignoredErr != nil && !pnet.IsDisconnectError(ignoredErr) {
+			mgr.logger.Error("close new backend connection failed", zap.Error(ignoredErr))
+		}
+		return err
+	}
+	mgr.backendIO.Store(newBackendIO)
```

Comment:
`mgr.backendIO` is already set in `getBackendIO`, so you should set it to nil when err != nil.

Reply:
Call `mgr.handshakeHandler.OnHandshake(mgr, mgr.ServerAddr(), nil, SrcNone)` after success.

Reply:
Fixed. I also called updateTraffic as Redirect do. PTAL if that's corret.